### PR TITLE
feat(suite-common): move analytics to common

### LINKED ISSUE #7596: Migrate analytics logic to common package
There is lot of analytics redux related logic in `package/suite`, move all these to `@suite-common/analytics`.

Output:
1. We are able to log events that are send somewhere. We need to solve how to distinguish between events from desktop and mobile.
2. Take different environment into account (we don't want to mix events from debug/dev/staging/production, how is this done in desktop Suite?)

### PR DESCRIPTION
**Pull request by juriczech**
## Description
Move analytics related redux logic so that we can reuse analytics in mobile app.

## Related Issue

Resolve https://github.com/trezor/trezor-suite/issues/7596

### CODE REVIEW

File: packages/suite/src/services/suite/metadata/DropboxProvider.ts [hunk 1]
```diff
@@ -3,7 +3,8 @@ import type { users } from 'dropbox';
 import { AbstractMetadataProvider } from '@suite-types/metadata';
 import { extractCredentialsFromAuthorizationFlow, getOauthReceiverUrl } from '@suite-utils/oauth';
 import { METADATA } from '@suite-actions/constants';
-import { getWeakRandomId } from '@trezor/utils';
+
+import { getWeakRandomId } from '@suite-common/suite-utils';
```
Comment (Nodonisko) [inline packages/suite/src/services/suite/metadata/DropboxProvider.ts]: This function should stay in `@trezor/utils`

File: packages/suite/src/utils/suite/__tests__/getWeakRandomId.test.ts [hunk 2]
```diff
@@ -1,4 +1,4 @@
-import { getWeakRandomId } from '../src/getWeakRandomId';
```
Comment (Nodonisko) [inline packages/suite/src/utils/suite/__tests__/getWeakRandomId.test.ts]: Why is this file moved from `@trezor/utils` to `@trezor/suite`?

File: packages/suite/src/utils/suite/analytics.ts [hunk 3]
```diff
@@ -29,12 +28,6 @@ export const redactTransactionIdFromAnchor = (anchor?: string) => {
     return anchor.startsWith(AccountTransactionBaseAnchor) ? AccountTransactionBaseAnchor : anchor;
 };
 
-// allow tracking only if user already confirmed data collection
-export const hasUserAllowedTracking = (
-    enabled: AnalyticsState['enabled'],
-    confirmed: AnalyticsState['confirmed'],
-) => !!confirmed && !!enabled;
-
 export const getSuiteReadyPayload = (state: AppState) => ({
```
Comment (Nodonisko) [inline packages/suite/src/utils/suite/analytics.ts:31]: We should have similar even in mobile app

File: suite-common/analytics/src/analyticsSlice.ts [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        builder.addCase(
+            extra.actionTypes.storageLoad,
+            (state, action: AnyAction) => action.payload.analytics || state,
+        );
+    },
+});
+
+// if instanceId does not exist yet (was not loaded from storage), create a new one
+export const selectAnalyticsInstanceId = (state: AnalyticsSliceRootState) =>
+    state.analytics.instanceId || getTrackingRandomId();
```
Comment (Nodonisko) [inline suite-common/analytics/src/analyticsSlice.ts]: this smells, I would rather avoid creating new id here.

File: suite-common/analytics/src/analyticsSlice.ts [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+// if user made choice, keep it, otherwise set it to true by default just to prefill the confirmation toggle
+export const selectIsAnalyticsEnabled = memoize((state: AnalyticsSliceRootState): boolean =>
```
Comment (Nodonisko) [inline suite-common/analytics/src/analyticsSlice.ts]: don't memoize, memoizing selectors with two simple if that returns primitive values doesn't make sense, memoizing here is probably slower than that two ifs itself

File: suite-common/analytics/src/analyticsSlice.ts [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    state.analytics.confirmed ? !!state.analytics.enabled : true,
+);
+
+export const selectIsAnalyticsConfirmed = (state: AnalyticsSliceRootState) =>
+    !!state.analytics.confirmed;
+
+// allow tracking only if user already confirmed data collection
+export const selectHasUserAllowedTracking = memoize(
```
Comment (Nodonisko) [inline suite-common/analytics/src/analyticsSlice.ts]: don't memoize

File: suite-native/analytics/package.json [hunk 7]
```diff
@@ -0,0 +1,20 @@
+{
+    "name": "@suite-native/analytics",
+    "version": "1.0.0",
+    "private": true,
+    "license": "See LICENSE.md in repo root",
+    "sideEffects": false,
+    "main": "src/index",
+    "scripts": {
+        "lint:js": "yarn g:eslint '**/*.{ts,tsx,js}'",
+        "test:unit": "jest -c ../../jest.config.base.js --passWithNoTests",
```
Comment (Nodonisko) [inline suite-native/analytics/package.json]: native config

File: suite-native/analytics/src/analyticsThunks.ts [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+    selectIsAnalyticsConfirmed,
+} from '@suite-common/analytics';
+import { createThunk } from '@suite-common/redux-utils';
+import { getTrackingRandomId } from '@trezor/analytics';
+import { getCommitHash, isDevelopEnv } from '@suite-native/config';
+
+import { analytics } from './analytics';
+import { EventType } from './constants';
+
+const ACTION_PREFIX = '@analytics';
```
Comment (Nodonisko) [inline suite-native/analytics/src/analyticsThunks.ts]: `@suite-native/analytics`

File: suite-common/analytics/src/analyticsSlice.ts [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+
+export type InitAnalyticsPayload = {
+    enabled: boolean;
+    confirmed: boolean;
+    instanceId: string;
+    sessionId: string;
+};
+
+export const prepareAnalyticsSlice = createSliceWithExtraDeps({
+    name: 'analytics',
```
Comment (Nodonisko) [inline suite-common/analytics/src/analyticsSlice.ts]: `@suite-common/analytics`

File: suite-native/app/src/initActions.ts [hunk 10]
```diff
@@ -0,0 +1,44 @@
+import { Alert } from 'react-native';
+
+import { createThunk } from '@suite-common/redux-utils';
+import { connectInitThunk } from '@suite-common/connect-init';
+import { initBlockchainThunk, reconnectBlockchainThunk } from '@suite-common/wallet-core';
+import { enabledNetworks } from '@suite-native/config';
+// import { initAnalytics } from '@trezor/connect-ui/libDev/src/utils/analytics';
```
Comment (Nodonisko) [inline suite-native/app/src/initActions.ts]: remove

File: suite-native/module-settings/src/screens/SettingsAnalyticsScreen.tsx [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+            dispatch(enableAnalyticsThunk());
+            return;
+        }
+        dispatch(disableAnalyticsThunk());
+    };
+
+    return (
+        <Screen>
+            <ScreenHeader title="Analytics" />
+            <Box flexDirection="row" justifyContent="space-between" marginVertical="medium">
```
Comment (Nodonisko) [inline suite-native/module-settings/src/screens/SettingsAnalyticsScreen.tsx]: We don't have component for this?

Anyway, this whole row should be clickable, it's horrible UX if you can click only that small switch.

**Review by Nodonisko — CHANGES_REQUESTED** [review]

File: suite-common/analytics/package.json [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+    "main": "src/index",
+    "scripts": {
+        "lint:js": "yarn g:eslint '**/*.{ts,tsx,js}'",
+        "test:unit": "jest -c ../../jest.config.base.js --passWithNoTests",
+        "type-check": "tsc --build"
+    },
+    "dependencies": {
+        "@reduxjs/toolkit": "1.8.3",
+        "@suite-common/redux-utils": "workspace:*",
+        "proxy-memoize": "^2.0.0"
```
Comment (Nodonisko) [inline suite-common/analytics/package.json]: could be removed I think

File: packages/analytics/src/analytics.ts [hunk 13]
```diff
@@ -1,8 +1,7 @@
 import { encodeDataToQueryString, getRandomId, getUrl, reportEvent } from './utils';
+import type { InitOptions, Event as AnalyticsEvenet, App } from './types';
```
Comment (tomasklim) [inline packages/analytics/src/analytics.ts]:
```suggestion
import type { InitOptions, Event as AnalyticsEvent, App } from './types';
```

File: packages/analytics/src/types.ts [hunk 14]
```diff
@@ -1,4 +1,4 @@
-export type App = 'suite' | 'connect';
+export type App = 'suite' | 'connect' | 'mobile';
```
Comment (tomasklim) [inline packages/analytics/src/types.ts:1]: What about using app `suite` with environment `mobile`?

File: suite-native/module-settings/src/screens/SettingsAnalyticsScreen.tsx [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+import { disableAnalyticsThunk, enableAnalyticsThunk } from '@suite-native/analytics';
+import { selectIsAnalyticsEnabled } from '@suite-common/analytics';
+
+import { TouchableSwitchRow } from '../components/TouchableSwitchRow';
+
+export const SettingsAnalyticsScreen = () => {
+    const dispatch = useDispatch();
+    const isAnalyticsEnabled = useSelector(selectIsAnalyticsEnabled);
+
+    const handleAnalyticsChange = (value: boolean) => {
```
Comment (tomasklim) [inline suite-native/module-settings/src/screens/SettingsAnalyticsScreen.tsx]: Better name for `value` variable

File: suite-common/analytics/src/analyticsSlice.ts [hunk 16]
```diff
@@ -0,0 +1,75 @@
+import { AnyAction, PayloadAction } from '@reduxjs/toolkit';
+import { memoize } from 'proxy-memoize';
+
+import { createSliceWithExtraDeps } from '@suite-common/redux-utils';
+
+type AnalyticsSliceState = {
+    sessionId?: string;
+    instanceId?: string;
+    enabled?: boolean;
+    confirmed?: boolean; // Confirmed through analytics consent?
```
Comment (tomasklim) [inline suite-common/analytics/src/analyticsSlice.ts]: yes

**Review by tomasklim — CHANGES_REQUESTED** [review]

File: packages/analytics/src/types.ts (hunk 14 above)
Comment (tomasklim) [reply packages/analytics/src/types.ts:1]: Hm, maybe it would make it too too complicated for analytics team. So we can probably delete `mobile` from `environment`?

File: packages/suite/src/actions/suite/__tests__/analyticsActions.test.ts [hunk 17]
```diff
@@ -1,146 +0,0 @@
...
-        };
-
-        expect(store.getActions()).toMatchObject([
-            unconfirmedInitAction,
-            unconfirmedInitAction,
-            unconfirmedInitAction,
-        ]);
-    });
-
-    it('analyticsActions.init() - enabled confirmed', () => {
```
Comment (tomasklim) [inline packages/suite/src/actions/suite/__tests__/analyticsActions.test.ts:95]: is this removed without any replacement? the unconfirmed logic is kinda crucial for us. We can lose critical data

File: packages/suite/src/services/suite/metadata/DropboxProvider.ts (hunk 1 above)
Comment (juriczech) [reply packages/suite/src/services/suite/metadata/DropboxProvider.ts]: Fixed

File: packages/suite/src/utils/suite/__tests__/getWeakRandomId.test.ts (hunk 2 above)
Comment (juriczech) [reply packages/suite/src/utils/suite/__tests__/getWeakRandomId.test.ts]: Fixed

File: suite-common/analytics/src/analyticsSlice.ts (hunk 4 above)
Comment (juriczech) [reply suite-common/analytics/src/analyticsSlice.ts]: Fixed

File: suite-common/analytics/src/analyticsSlice.ts (hunk 6 above)
Comment (juriczech) [reply suite-common/analytics/src/analyticsSlice.ts]: Removed

File: suite-common/analytics/src/analyticsSlice.ts (hunk 5 above)
Comment (juriczech) [reply suite-common/analytics/src/analyticsSlice.ts]: In the end I composed this from other selector as well so I kept the memoize

File: suite-native/analytics/package.json (hunk 7 above)
Comment (juriczech) [reply suite-native/analytics/package.json]: Fixed

File: suite-native/analytics/src/analyticsThunks.ts (hunk 8 above)
Comment (juriczech) [reply suite-native/analytics/src/analyticsThunks.ts]: Fixed

File: suite-common/analytics/src/analyticsSlice.ts (hunk 16 above)
Comment (juriczech) [reply suite-common/analytics/src/analyticsSlice.ts]: This was just comment to explain what it means - there was comment in previous reducer as well. I guess I can remove it as it's kind of self explanatory

Comment (juriczech) [reply suite-common/analytics/src/analyticsSlice.ts]: https://github.com/trezor/trezor-suite/pull/7622/commits/48d2e6372febcff336d648844ed47bc0f8784ca6

File: suite-native/module-settings/src/screens/SettingsAnalyticsScreen.tsx (hunk 15 above)
Comment (juriczech) [reply suite-native/module-settings/src/screens/SettingsAnalyticsScreen.tsx]: https://github.com/trezor/trezor-suite/pull/7622/commits/48d2e6372febcff336d648844ed47bc0f8784ca6

File: packages/analytics/src/analytics.ts (hunk 13 above)
Comment (juriczech) [reply packages/analytics/src/analytics.ts]: https://github.com/trezor/trezor-suite/pull/7622/commits/48d2e6372febcff336d648844ed47bc0f8784ca6

File: packages/suite/src/actions/suite/__tests__/analyticsActions.test.ts (hunk 17 above)
Comment (juriczech) [reply packages/suite/src/actions/suite/__tests__/analyticsActions.test.ts:95]: https://github.com/trezor/trezor-suite/pull/7622/commits/9699a35bef2f083ee0118dad7e93d3a7474def8b

File: packages/analytics/src/types.ts (hunk 14 above)
Comment (juriczech) [reply packages/analytics/src/types.ts:1]: https://github.com/trezor/trezor-suite/pull/7622/commits/9699a35bef2f083ee0118dad7e93d3a7474def8b

File: suite-native/module-settings/src/screens/SettingsAnalyticsScreen.tsx (hunk 11 above)
Comment (juriczech) [reply suite-native/module-settings/src/screens/SettingsAnalyticsScreen.tsx]: fixed

File: packages/suite/src/actions/suite/__tests__/analyticsActions.test.ts (hunk 17 above)
Comment (juriczech) [reply packages/suite/src/actions/suite/__tests__/analyticsActions.test.ts:95]: I thought we won't be testing slice reducers as this is not as complicated on logic as typescript checks whether types are correct and then it's just kind of "documentation" to have these tests. Since the API changed and I couldn't just move these tests to common, I figured I'll get rid of this. 

But if you think it should be tested as a critical part of app, I can put them back 😉

Comment (juriczech) [reply packages/suite/src/actions/suite/__tests__/analyticsActions.test.ts:95]: https://github.com/trezor/trezor-suite/pull/7622/commits/16bc07eb8cf54f4b869d6e7028c4693fd3d63e4a

File: packages/analytics/src/types.ts [hunk 18]
```diff
@@ -1,6 +1,6 @@
...
 
-export type Environment = 'desktop' | 'web' | 'mobile' | '';
+export type Environment = 'desktop' | 'web' | '';
```
Comment (Nodonisko) [inline packages/analytics/src/types.ts]: why is there empty string?

Comment (juriczech) [reply packages/analytics/src/types.ts]: https://github.com/trezor/trezor-suite/pull/7622/commits/54d904a8ba8133fa6fe8abfb21be6c6fd7bb4820

File: packages/suite/src/actions/suite/initAction.ts [hunk 19]
```diff
@@ -28,7 +27,7 @@ export const init = () => async (dispatch: Dispatch, getState: GetState) => {
     // right after storage is loaded, we might start:
 
     // 1. init analytics
-    dispatch(analyticsActions.init(analytics));
```
Comment (Nodonisko) [inline packages/suite/src/actions/suite/initAction.ts:31]: is ok that you removed that param?

File: suite-common/analytics/src/analyticsReducer.ts [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+);
+
+// if instanceId does not exist yet (was not loaded from storage), create a new one
+export const selectAnalyticsInstanceId = (state: AnalyticsRootState) => state.analytics.instanceId;
+
+export const selectIsAnalyticsConfirmed = (state: AnalyticsRootState) =>
+    !!state.analytics.confirmed;
+
+// if user made choice, keep it, otherwise set it to true by default just to prefill the confirmation toggle
+export const selectIsAnalyticsEnabled = memoize((state: AnalyticsRootState): boolean => {
```
Comment (Nodonisko) [inline suite-common/analytics/src/analyticsReducer.ts]: memoize not neccessary, it's very simple operation and return value is primitive type

File: suite-common/analytics/package.json (hunk 12 above)
Comment (Nodonisko) [reply suite-common/analytics/package.json]: still here, could be removed

File: suite-common/suite-types/src/environment.ts [hunk 21]
```diff
@@ -1 +1 @@
-export type EnvironmentType = 'web' | 'desktop' | 'mobile' | '';
+export type EnvironmentType = 'web' | 'desktop';
```
Comment (Nodonisko) [inline suite-common/suite-types/src/environment.ts]: This type in exactly same shape is also in `@suite-common/analytics`, I think we should remove it in one of places.

File: suite-native/analytics/src/analyticsThunks.ts [hunk 22]
```diff
@@ -0,0 +1,60 @@
...
+
+export const disableAnalyticsThunk = createThunk(
+    `${ACTION_PREFIX}/disableAnalyticsThunk`,
+    (_, { dispatch }) => {
+        analytics.report({ type: EventType.SettingsAnalytics, payload: { value: false } }, true);
+        dispatch(analyticsActions.disableAnalytics());
+    },
+);
+
+export const initAnalytics = createThunk(`${ACTION_PREFIX}/init`, (_, { dispatch, getState }) => {
```
Comment (Nodonisko) [inline suite-native/analytics/src/analyticsThunks.ts]: `initAnalyticsThunk`

**Review by Nodonisko — CHANGES_REQUESTED** [review]

File: suite-common/analytics/src/analyticsReducer.ts [hunk 23]
```diff
@@ -0,0 +1,66 @@
...
+    },
+);
+
+// if instanceId does not exist yet (was not loaded from storage), create a new one
+export const selectAnalyticsInstanceId = (state: AnalyticsRootState) => state.analytics.instanceId;
+
+export const selectIsAnalyticsConfirmed = (state: AnalyticsRootState) =>
+    !!state.analytics.confirmed;
+
+// if user made choice, keep it, otherwise set it to true by default just to prefill the confirmation toggle
```
Comment (tomasklim) [inline suite-common/analytics/src/analyticsReducer.ts]: This comment is outdated now?

File: packages/suite/src/actions/suite/initAction.ts (hunk 19 above)
Comment (juriczech) [reply packages/suite/src/actions/suite/initAction.ts:31]: Yes. See `analyticsActions.ts` where I updated thunk params

File: suite-common/suite-types/src/environment.ts (hunk 21 above)
Comment (juriczech) [reply suite-common/suite-types/src/environment.ts]: No it's not. Maybe you checked wrong commit? `@suite-common/analytics` currently only has actions and reducer

File: suite-common/analytics/src/analyticsReducer.ts (hunk 23 above)
Comment (juriczech) [reply suite-common/analytics/src/analyticsReducer.ts]: Good point 👌
https://github.com/trezor/trezor-suite/pull/7622/commits/e5862a187c231e0b5bd505afe0e2f76fd408f177

File: suite-native/analytics/src/analyticsThunks.ts (hunk 22 above)
Comment (juriczech) [reply suite-native/analytics/src/analyticsThunks.ts]: https://github.com/trezor/trezor-suite/pull/7622/commits/e5862a187c231e0b5bd505afe0e2f76fd408f177

File: suite-common/analytics/src/analyticsReducer.ts (hunk 20 above)
Comment (juriczech) [reply suite-common/analytics/src/analyticsReducer.ts]: https://github.com/trezor/trezor-suite/pull/7622/commits/e5862a187c231e0b5bd505afe0e2f76fd408f177

File: packages/suite/src/actions/suite/initAction.ts (hunk 19 above)
Comment (juriczech) [reply packages/suite/src/actions/suite/initAction.ts:31]: Previously, state was passed through parameter and the thunk was picking properties from previous state. I decided it's not necessary when `getState` is available via thunk props. Consulted this with @tomasklim and he verified that there was probably no reason for this and could be refactored. It's better like this IMO.

File: suite-common/suite-types/src/environment.ts (hunk 21 above)
Comment (Nodonisko) [reply suite-common/suite-types/src/environment.ts]: Exactly same type is in file `packages/analytics/src/types.ts`

File: suite-common/wallet-core/src/transactions/transactionsActions.ts [hunk 24]
```diff
@@ -4,7 +4,7 @@ import { Account, WalletAccountTransaction } from '@suite-common/wallet-types';
 import { AccountTransaction } from '@trezor/connect';
 import { enhanceTransaction } from '@suite-common/wallet-utils';
 
-export const modulePrefix = '@common/wallet-core/transactions';
+export const modulePrefix = '@suite-common/wallet-core/transactions';
```
Comment (Nodonisko) [inline suite-common/wallet-core/src/transactions/transactionsActions.ts]: Please avoid these unrelated changes, it makes checking PR harder. Also shortening of prefixes for actions to just `@common` or `@native`is fine and it's like that everywhere.

**Review by Nodonisko — CHANGES_REQUESTED** [review]

File: suite-common/wallet-core/src/transactions/transactionsActions.ts (hunk 24 above)
Comment (juriczech) [reply suite-common/wallet-core/src/transactions/transactionsActions.ts]: https://github.com/trezor/trezor-suite/pull/7622/commits/c7f5d2843b15663e8216eaf6606346df93acc59e

File: suite-common/suite-types/src/environment.ts (hunk 21 above)
Comment (juriczech) [reply suite-common/suite-types/src/environment.ts]: https://github.com/trezor/trezor-suite/pull/7622/commits/c7f5d2843b15663e8216eaf6606346df93acc59e

**Review by Nodonisko — APPROVED** [review]

**Review by tomasklim — APPROVED** [review]
